### PR TITLE
[breaking batch] Remove some unnecessary indirection from AST structures

### PR DESCRIPTION
**Pull request by petrochenkov**
cc #31487
plugin-[breaking-change]

The AST part of https://github.com/rust-lang/rust/pull/30087

r? @Manishearth

### CODE REVIEW

**Comment by Manishearth** [comment]
@bors r+

LGTM. Not sure why Annotatable still uses P, but that's okay.

**Comment by petrochenkov** [comment]
Huh, it seems like without prioritizing two PRs approved roughly at the same time have to be extremely lucky to end up in the same nightly.

**Comment by Manishearth** [comment]
@bors p=10

shit sorry, I forgot

File: src/libsyntax/ptr.rs [hunk 1]
```diff
@@ -65,6 +65,10 @@ impl<T: 'static> P<T> {
     {
         f(*self.ptr)
     }
+    /// Equivalent to and_then(|x| x)
+    pub fn unwrap(self) -> T {
```
Comment (eddyb) [inline src/libsyntax/ptr.rs]: Wow I'm years late to this but see https://github.com/rust-lang/rust/pull/54277#discussion_r257181754

**Pull request merged**